### PR TITLE
Port commits from 3.1.x onto master (resolves #683, resolves #686, resolves #690)

### LINKED ISSUE #683: logToMaster() doesn't work
This is an example from the docs. It does not print the logToMaster() message, neither on master nor on 3.1.x.

```
from toil.job import Job

class HelloWorld(Job):
    def __init__(self, message):
        Job.__init__(self,  memory="2G", cores=2, disk="3G")
        self.message = message

    def run(self, fileStore):
        fileStore.logToMaster("Hello, world!, I have a message: %s"
                              % self.message)

if __name__=="__main__":
    options = Job.Runner.getDefaultOptions("./toilWorkflowRun")
    options.logLevel = "INFO"
    Job.Runner.startToil(HelloWorld("woot"), options)
```

### LINKED ISSUE #686: Add option for retaining files in work directory (--workDir)
Allow users a way to retain temporary files generated during a toil run. Should probably be a new option (not --clean) since it will be very detrimental to toil's performance and only make sense in a debugging situation.

### LINKED ISSUE #690: A failing root job causes EOFError
Discovered while testing #686. To replicate:
```python
options = Job.Runner.getDefaultOptions(self._getTestJobStorePath())
options.workDir = self.testDir
options.clean = "always"
options.cleanWorkDir = "onSuccess"
A = Job.wrapJobFn(tempFileTestErrorJob)
Job.Runner.startToil(A, options)
```
Where tempFileTestErrorJob is
```python
def tempFileTestErrorJob(job):
    with open(job.fileStore.getLocalTempFile(), "w") as f:
        f.write("test file retention")
    raise RuntimeError()
```
The EOF error occurs when toil tries to unpickle the root job's return value:

        #Parse out the return value from the root job
        with jobStore.readSharedFileStream("rootJobReturnValue") as fH:
            jobStoreFileID = fH.read()
        with jobStore.readFileStream(jobStoreFileID) as fH:
            rootJobReturnValue = cPickle.load(fH)  #EOF here

We should probably catch this and instead raise a toil.leader.FailedJobsException as documented

### PR DESCRIPTION
**Pull request by cket**
Cherry picked already merged commits from 3.1.x branch. Resolves a logToMaster issue, EOF error upon root job failure, and adds support for optionally retaining the workDir on workers.

### CODE REVIEW

File: src/toil/batchSystems/abstractBatchSystem.py [hunk 1]
```diff
@@ -160,16 +162,17 @@ def supportsWorkerCleanup():
...
         :param collections.namedtuple workerCleanupInfo: A named tuple consisting of all the
         relevant information for cleaning up the worker.
-        '''
+        """
         assert workerCleanupInfo.__class__.__name__ == 'WorkerCleanupInfo'
         workflowDir = getToilWorkflowDir(workerCleanupInfo.workflowID, workerCleanupInfo.workDir)
-        # TODO: logic here for cleanup flag
-        shutil.rmtree(workflowDir)
-
+        dirIsEmpty = os.listdir(workflowDir) == []
```
Comment (cket) [inline src/toil/batchSystems/abstractBatchSystem.py]: @hannes-ucsc I checked and did not find a method that simplifies checking if a directory is empty

**Comment by cket** [comment]
@hannes-ucsc I will merge this PR later today per your instructions yesterday

File: src/toil/batchSystems/abstractBatchSystem.py (hunk 1 above)
Comment (hannes-ucsc) [inline src/toil/batchSystems/abstractBatchSystem.py]: Yes, I was wrong. `os.stat(path).n_link >2` only works to check if a directory contains sub-directories.

**Comment by hannes-ucsc** [comment]
> I will merge this

SGTM

**Pull request merged**